### PR TITLE
Enable coupling of time derivatives of NL vars into Aux system

### PR DESCRIPTION
**Pull request by andrsd**
The computation of time derivatives of non-linear variables can be done
ahead, since u_n{+1} and u{n} are known. Then, we can go update the aux
variables and we can allow the time derivatives of NL vars there.
Finally, NL vars in kernels, BC, etc. can use the time derivatives of
aux and NL vars, since the current solution vectors are up to date.

Coupling time derivatives of aux variables into aux system is not
allowed and will throw an error.

Refs #442

### CODE REVIEW

**Comment by friedmud** [comment]
:+1:

**Comment by YaqiWang** [comment]
IMHO coupling time derivatives of aux variables into aux system can also be easily done. No need to let that moose error hang there.

**Comment by andrsd** [comment]
I agree it can be done, but I do not think that it can be done **easily**.

**Comment by YaqiWang** [comment]
The only problem I see is that we will need to evaluate u_dot for aux variables right after aux kernel compute, which can be done in AuxKernel. We will need to have time integrators provide a utility function like
```
computeTimeDerivatives(const VariableValue & u, const VariableValue & u_old, VariableValue & u_dot);
```
Let AuxKernel get a reference of time integrator is not a terrible idea to me.

**Comment by andrsd** [comment]
The problem I see is: if aux `a` needs `b_dot`, then I need to compute the value of `b` first, then I can compute `b_dot`, then I can start evaluating `a`. So, we would need this interleaved execution/computation in the aux-system. Time integrators are currently working with the whole solution vectors, so in order to do this, we would need some non-trivial work done. That's why I do not think it is **easily** possible.

**Comment by YaqiWang** [comment]
When `a` calls `coupledDot(b)`, it creates the dependency between `a` and `b`. So `b` will be evaluated first and `a` will see `b_dot`. On the aux system, the whole vector version of `computeTimeDerivatives()` will not be called, instead the element version I provided above should be called. We just pass in the references of the variable obtained in `MooseVariable`. Did I miss something?

**Comment by andrsd** [comment]
I was not saying there is dependency issue, but I was rather showing the interleaving issue. In your suggestion, we would have to compute `u_dot` point-wise (for linear lagrange), but our system is not doing it.
Again, I am not saying it is not possible, but that is not that easy. Unless you do an ugly and dirty hack to do it easily ;-)

**Comment by YaqiWang** [comment]
This actually can used to solve one of my very old problems ;-)
I have a equation for u like
```
u_t = -c u + v
```
where `v` is the primal variable and `u` can be treated as an aux variable because there is not spatial derivative terms in its equation. But currently I have no way to obtain information in time integrator to do the `u` evaluation correctly. It is kind of like the aux variable is coupling his own dot.

If we let the time integrator provide an element version, then I can somehow update u correctly in my aux kernel no matter what time integrator is being used. And I can make the time integration of my aux variable always consistent with the integrator used for primal system.

File: test/include/auxkernels/DotCouplingAux.h [hunk 1]
```diff
@@ -0,0 +1,26 @@
+#ifndef DOTCOUPLINGAUX_H
```
Comment (permcody) [inline test/include/auxkernels/DotCouplingAux.h]: Missing MOOSE header

File: test/src/auxkernels/DotCouplingAux.C [hunk 2]
```diff
@@ -0,0 +1,26 @@
+#include "DotCouplingAux.h"
```
Comment (permcody) [inline test/src/auxkernels/DotCouplingAux.C]: Missing MOOSE header

**Pull request merged**